evap/staff/importers/user.py

```diff
@@ -0,0 +1,384 @@
+import functools
+import operator
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Dict, Iterable, Sequence, Set, Union
+
+from django.core.exceptions import ValidationError
+from django.db import transaction
+from django.db.models import Q
+from django.utils.html import format_html
+from django.utils.translation import gettext as _
+
+from evap.evaluation.models import UserProfile
+from evap.evaluation.tools import clean_email
+from evap.staff.tools import create_user_list_html_string_for_message
+
+from .base import (
+    Checker,
+    ConvertExceptionsToMessages,
+    ExcelFileLocation,
+    ExcelFileRowMapper,
+    FirstLocationAndCountTracker,
+    ImporterLog,
+    ImporterLogEntry,
+    InputRow,
+    RowCheckerMixin,
+)
+
+
+@dataclass(eq=True, frozen=True)
+class UserData:
+    """Holds information about a user from the import file (may have to be created)"""
+
+    first_name: str
+    last_name: str
+    title: str
+    email: str
+
+    @staticmethod
+    def bulk_update_fields():
+        """Fields passed to bulk_update when updating existing users with new UserData"""
+        return ["first_name", "last_name", "title", "email", "is_active"]
+
+    def __init__(self, first_name: str, last_name: str, title: str, email: str):
+        # object.__setattr__ is needed to initialize instances of frozen dataclasses
+        object.__setattr__(self, "first_name", first_name.strip())
+        object.__setattr__(self, "last_name", last_name.strip())
+        object.__setattr__(self, "title", title.strip())
+        object.__setattr__(self, "email", clean_email(email))
+
+    def apply_to_and_make_active(self, user_profile: UserProfile):
+        """Intended to update existing UserProfile entries from the database. email is not touched"""
+        user_profile.first_name = self.first_name
+        user_profile.last_name = self.last_name
+        user_profile.title = self.title
+        user_profile.is_active = True
+
+    def get_user_profile_object(self):
+        """Create a new UserProfile object with the same data. Used for validation and bulk insertion"""
+        obj = UserProfile(email=self.email)
+        self.apply_to_and_make_active(obj)
+        return obj
+
+    def validate(self):
+        user = self.get_user_profile_object()
+
+        # User might already exist in the database. In this case, we will later update the existing user.
+        # blank password would trigger an error here although its fine for us.
+        user.full_clean(exclude=["password"], validate_unique=False)
+
+
+@dataclass
+class UserInputRow(InputRow):
+    """Raw representation of a semantic user importer row, independent on the import format (xls, csv, ...)"""
+
+    column_count = 4
+
+    location: ExcelFileLocation
+
+    title: str
+    first_name: str
+    last_name: str
+    email: str
+
+    def __init__(self, location: ExcelFileLocation, cells: Sequence[str]):
```

Comment:
I considered the fact that the attributes and cell values line up to be accidental here, so I decided I did not want to rely on that. I guess this is up to personal taste? Considering that the current method is 4 lines only, and just the declaration of the abstract method would also be 4 lines, I don't see how it could be any shorter.
Would you also be ok with the current code?

Reply:
What really led me to this suggestion was `EnrollmentInputRow.__init__` which assigns 12 cells out of order, which is harder to grasp.

You are right, whether we want to have the dataclass declaration be exactly equivalent to the order of cells is a elegant-or-awful debate. I think that it would be a compact and declarative way of communicating the order, and since the `InputRow` classes don't do much more anyway, it would not be too complicated either.

Of course, one can list the same arguments for the other side and the current version isn't doing any actual harm, so I am not strongly for or against either side, the current code would also be fine for me

Reply:
Thought about it again, agreed, applied your version.